src/libutil/compression.cc

```diff
@@ -40,20 +38,26 @@ struct ArchiveDecompressionSource : Source
 {
     std::unique_ptr<TarArchive> archive = 0;
     Source & src;
-    ArchiveDecompressionSource(Source & src) : src(src) {}
+    std::string compressionMethod;
+    ArchiveDecompressionSource(Source & src, std::string compressionMethod = "")
```

Comment:
(nit)

```suggestion
    std::optional<std::string> compressionMethod;
    ArchiveDecompressionSource(Source & src, std::optional<std::string> compressionMethod = "")
```

Doesn't change much but makes the intent slightly more explicit

Reply:
Huh, I've changed to using `std::optional` and I'm seeing a very curious bug. Here method is passed as a const ref:

```
std::unique_ptr<FinishSink> makeDecompressionSink(const std::string & method, Sink & nextSink)
{
    if (method == "none" || method == "")
        return std::make_unique<NoneSink>(nextSink);
    else if (method == "br")
        return std::make_unique<BrotliDecompressionSink>(nextSink);
    else
        return sourceToSink([&](Source & source) {
            auto decompressionSource = std::make_unique<ArchiveDecompressionSource>(source, method);
            decompressionSource->drainInto(nextSink);
        });
}
```

However, there's certainly some UB going on here because a const ref `method` becomes an empty string here:
```
return sourceToSink([&](Source & source) {
            // here on my machine exception is actually thrown. Though this should never happen 
            // (by simple inspection) in a well-defined program. 
            if (method.empty()) throw std::runtime_error{"VERY BAD UB AND DATARACES"};
            auto decompressionSource = std::make_unique<ArchiveDecompressionSource>(source, method);
            decompressionSource->drainInto(nextSink);
        });
```

But this is a courotine created here: `src/libstore/filetransfer.cc:163`. `encoding` is passed by reference, but it looks
like another thread races on the same variable and sets it to an empty string here: `src/libstore/filetransfer.cc:208`. 

Isn't nix built with thread sanitizer to actually catch such blatant UB? Not only is this a datarace (maybe a race on the same thread between different tasks/coroutines), but also a modification of a supposedly const ref. 

The correct thing to do here is obviosly take the string by value. But such bugs are really better diagnosed with sanitizers

Reply:
Would it maybe be beneficial to open an issue/rfc to introduce tests under address/memory/ub/thread sanitizers. I would not be surprised it this is happening in other places as well...

Reply:
It would be great if someone could reproduce this issue to make sure I'm not braindead and not going crazy

Reply:
I've pinned down what's going on. There was either a dangling reference/race. Capturing `method` by value into the lambda passed to the coroutine resolves the issue:

```cpp
        return sourceToSink([method, &nextSink](Source & source) {
            auto decompressionSource = std::make_unique<ArchiveDecompressionSource>(source, method);
            decompressionSource->drainInto(nextSink);
        });
```

Reply:
Nice investigation (and sorry for the headaches it caused you).

Regarding 

> Would it maybe be beneficial to open an issue/rfc to introduce tests under address/memory/ub/thread sanitizers. I would not be surprised it this is happening in other places as well...

100x yes! I opened https://github.com/NixOS/nix/pull/10144 for that last month, but didn't have time to finish it. Let me know if you're interested in picking it up :)